api/billingd/filrewards/filrewards.go

```diff
@@ -0,0 +1,217 @@
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.DBURI))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.DBName)
+	col := db.Collection(config.CollectionName)
+	_, err = col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys:    bson.D{primitive.E{Key: "key", Value: 1}, primitive.E{Key: "event", Value: 1}},
+			Options: options.Index().SetUnique(true).SetSparse(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "key", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "event", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating indexes: %v", err)
+	}
+	// Populate cache.
+	opts := options.Find()
+	cursor, err := col.Find(ctx, bson.M{}, opts)
+	if err != nil {
+		return nil, fmt.Errorf("querying RewardRecords to populate cache: %s", err)
+	}
+	defer cursor.Close(ctx)
+	cache := make(map[string]map[Event]struct{})
+	for cursor.Next(ctx) {
+		var rec RewardRecord
+		if err := cursor.Decode(&rec); err != nil {
+			return nil, fmt.Errorf("decoding RewardRecord while building cache: %v", err)
+		}
+		ensureKeyEventCache(cache, rec.Key)
+		cache[rec.Key][rec.Event] = struct{}{}
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("iterating cursor while building cache: %v", err)
+	}
+	return &FilRewards{
+		col:          col,
+		rewardsCache: cache,
+	}, nil
+}
+
+func (f *FilRewards) ProcessEvent(ctx context.Context, key string, accountType mdb.AccountType, event analytics.Event) (*RewardRecord, error) {
+	rewardEvent := Unspecified
+	switch event {
+	case analytics.KeyAccountCreated:
+		rewardEvent = FirstKeyAccountCreated
+	case analytics.KeyUserCreated:
+		rewardEvent = FirstKeyUserCreated
+	case analytics.OrgCreated:
+		rewardEvent = FirstOrgCreated
+	case analytics.BillingSetup:
+		rewardEvent = InitialBillingSetup
+	case analytics.BucketCreated:
+		rewardEvent = FirstBucketCreated
+	case analytics.BucketArchiveCreated:
+		rewardEvent = FirstBucketArchiveCreated
+	case analytics.MailboxCreated:
+		rewardEvent = FirstMailboxCreated
+	case analytics.ThreadDbCreated:
+		rewardEvent = FirstThreadDbCreated
+	}
+	if rewardEvent == Unspecified {
+		return nil, nil
+	}
+
+	ensureKeyEventCache(f.rewardsCache, key)
+
+	if _, exists := f.rewardsCache[key][rewardEvent]; exists {
+		// This reward is already granted so bail.
+		return nil, nil
+	}
+
+	rec := RewardRecord{
+		Key:               key,
+		AccountType:       accountType,
+		Event:             rewardEvent,
+		Factor:            eventRewards[rewardEvent].factor,
+		BaseAttoFILReward: baseAttoFILReward,
+		CreatedAt:         time.Now(),
+	}
+
+	if _, err := f.col.InsertOne(ctx, rec); err != nil {
+		return nil, fmt.Errorf("inserting RewardRecord: %v", err)
+	}
+
+	f.rewardsCache[key][rewardEvent] = struct{}{}
+
+	// ToDo: Notify the account owner.
```

Comment:
Important `ToDo`.

Reply:
Email or something?

Reply:
you should just create a new analytics event with the reward info `Track()`. emails can then be linked up from there.

Reply:
Do you mean to add the reward info as metadata to the segment event? Cool, so you can trigger emails from within segment, or? Just want to understand how that will all work.

Reply:
yeah, we can trigger emails there. best to create a specific event type for each reward, then yeah, include any useful metadata (e.g. fil rewarded) and we can automate the email from that